srml/grandpa/src/lib.rs

```diff
@@ -302,79 +242,61 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// Deposit one of this module's logs.
-	fn deposit_log(log: Log<T>) {
-		<system::Module<T>>::deposit_log(<T as Trait>::Log::from(log).into());
+	fn deposit_log(log: Signal<T::BlockNumber>) {
+		let log: DigestItem<T::Hash> = DigestItem::Consensus(GRANDPA_ENGINE_ID, log.encode());
+		<system::Module<T>>::deposit_log(log.into());
 	}
 }
 
-impl<T: Trait> Module<T> where AuthorityId: core::convert::From<<T as Trait>::SessionKey> {
-	/// See if the digest contains any standard scheduled change.
-	pub fn scrape_digest_change(log: &Log<T>)
+impl<T: Trait> Module<T> {
+	pub fn grandpa_log(digest: &T::Digest) -> Option<Signal<T::BlockNumber>> {
+		let id = OpaqueDigestItemId::Consensus(&GRANDPA_ENGINE_ID);
+		digest.convert_first(|l| l.try_to::<Signal<T::BlockNumber>>(id))
+	}
+
+	pub fn pending_change(digest: &T::Digest)
 		-> Option<ScheduledChange<T::BlockNumber>>
 	{
-		<Log<T> as GrandpaChangeSignal<T::BlockNumber>>::as_signal(log)
+		Self::grandpa_log(digest).and_then(|signal| signal.try_into_change())
 	}
 
-	/// See if the digest contains any forced scheduled change.
-	pub fn scrape_digest_forced_change(log: &Log<T>)
+	pub fn forced_change(digest: &T::Digest)
 		-> Option<(T::BlockNumber, ScheduledChange<T::BlockNumber>)>
 	{
-		<Log<T> as GrandpaChangeSignal<T::BlockNumber>>::as_forced_signal(log)
+		Self::grandpa_log(digest).and_then(|signal| signal.try_into_forced_change())
 	}
 }
 
-/// Helper for authorities being synchronized with the general session authorities.
-///
-/// This is not the only way to manage an authority set for GRANDPA, but it is
-/// a convenient one. When this is used, no other mechanism for altering authority
-/// sets should be.
-pub struct SyncedAuthorities<T>(::rstd::marker::PhantomData<T>);
-
-// FIXME: remove when https://github.com/rust-lang/rust/issues/26925 is fixed
-impl<T> Default for SyncedAuthorities<T> {
-	fn default() -> Self {
-		SyncedAuthorities(::rstd::marker::PhantomData)
-	}
-}
-
-impl<X, T> session::OnSessionChange<X> for SyncedAuthorities<T> where
-	T: Trait + consensus::Trait<SessionKey=<T as Trait>::SessionKey>,
-	<T as consensus::Trait>::Log: From<consensus::RawLog<<T as Trait>::SessionKey>>
-{
-	fn on_session_change(_: X, _: bool) {
-		use primitives::traits::Zero;
-
-		let next_authorities = <consensus::Module<T>>::authorities()
-			.into_iter()
-			.map(|key| (key, 1)) // evenly-weighted.
-			.collect::<Vec<(<T as Trait>::SessionKey, u64)>>();
-
+impl<T: Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
+	type Key = AuthorityId;
+	fn on_new_session<'a, I: 'a>(changed: bool, validators: I)
```

Comment:
The previous approach was not to tightly couple Grandpa's authority-set management with the `Session` module. That's the purpose of `SyncedAuthorities` which should be the implementer of `OnSessionChange`. This is more flexible API, that we also use to set up `CuratedGrandpa` for `Sudo` management of authorities.

Reply:
grandpa itself is not coupled with session - it can be used entirely independently. if we still want to bother with a curated grandpa set, then it's just a case of making the driving it either through the `OneSessionHandler` API or directly through the modules `pub` members. i just didn't see the point in introducing an extra type with a whole load of trait bounds when we can impl directly for the module.

if you mean the fact that this introduces a dependency on session from grandpa, then it's easy enough to decouple that: `OneSessionHandler` doesn't need to be in the `session` crate - it could reasonably be moved to the `srml_support::traits` where a lot of support traits already live.